File: js/src/backends/link-cli-process.js
/**
 * link-cli process management for links-queue.
 *
 * This module handles the lifecycle of link-cli child processes,
 * including spawning, communication via stdin/stdout, and graceful shutdown.
 *
 * @module backends/link-cli-process
 *
 * @see https://github.com/link-foundation/link-cli
 */

import { spawn } from 'node:child_process';
import { EventEmitter } from 'node:events';
import { setTimeout, clearTimeout } from 'node:timers';

/**
 * Process state enum.
 * @readonly
 * @enum {string}
 */
export const ProcessState = {
  /** Process not yet started */
  STOPPED: 'stopped',
  /** Process is starting up */
  STARTING: 'starting',
  /** Process is running and ready */
  RUNNING: 'running',
  /** Process is shutting down */
  STOPPING: 'stopping',
  /** Process has crashed */
  CRASHED: 'crashed',
};

/**
 * Default options for LinkCliProcess.
 * @type {Object}
 */
const DEFAULT_OPTIONS = {
  /** Path to link-cli binary */
  binaryPath: 'clink',
  /** Path to database file */
  dbPath: 'db.links',
  /** Whether to auto-restart on crash */
  autoRestart: true,
  /** Maximum restart attempts */
  maxRestarts: 3,
  /** Delay between restart attempts in ms */
  restartDelay: 1000,
  /** Timeout for operations in ms */
  operationTimeout: 30000,
  /** Enable verbose/trace output from link-cli */
  trace: false,
};

/**
 * Manages a link-cli child process.
 *
 * Handles spawning, communication, and lifecycle management of the
 * link-cli process. Each operation spawns a new process and waits
 * for completion (link-cli operates in command mode, not interactive).
 *
 * @extends EventEmitter
 *
 * @example
 * const process = new LinkCliProcess({
 *   dbPath: './data/db.links',
 *   binaryPath: 'clink'
 * });
 *
 * // Execute a query
 * const result = await process.execute('() ((1 1))');
 *
 * @fires LinkCliProcess#ready - When process is ready for operations
 * @fires LinkCliProcess#error - When an error occurs
 * @fires LinkCliProcess#exit - When process exits
 */
export class LinkCliProcess extends EventEmitter {
  /**
   * Creates a new LinkCliProcess instance.
   *
   * @param {Object} [options={}] - Configuration options
   * @param {string} [options.binaryPath='clink'] - Path to link-cli binary
   * @param {string} [options.dbPath='db.links'] - Path to database file
   * @param {boolean} [options.autoRestart=true] - Auto-restart on crash
   * @param {number} [options.maxRestarts=3] - Maximum restart attempts
   * @param {number} [options.restartDelay=1000] - Delay between restarts (ms)
   * @param {number} [options.operationTimeout=30000] - Operation timeout (ms)
   * @param {boolean} [options.trace=false] - Enable trace output
   */
  constructor(options = {}) {
    super();

    /** @private */
    this._options = { ...DEFAULT_OPTIONS, ...options };

    /** @private */
    this._state = ProcessState.STOPPED;

    /** @private */
    this._restartCount = 0;
  }

  /**
   * Gets the current process state.
   * @returns {string} Current state
   */
  get state() {
    return this._state;
  }

  /**
   * Gets the database path.
   * @returns {string} Database path
   */
  get dbPath() {
    return this._options.dbPath;
  }

  /**
   * Gets the binary path.
   * @returns {string} Binary path
   */
  get binaryPath() {
    return this._options.binaryPath;
  }

  /**
   * Checks if link-cli binary is available.
   *
   * @returns {Promise<boolean>} True if binary is available
   */
  // eslint-disable-next-line require-await
  async checkBinaryAvailable() {
    return new Promise((resolve) => {
      const proc = spawn(this._options.binaryPath, ['--help'], {
        stdio: ['ignore', 'pipe', 'pipe'],
        timeout: 5000,
      });

      proc.on('error', () => resolve(false));
      proc.on('close', (code) => resolve(code === 0));
    });
  }

  /**
   * Executes a link-cli query and returns the result.
   *
   * link-cli operates in command mode - each operation spawns a new
   * process, executes the query, and returns the result.
   *
   * @param {string} query - The Links Notation query to execute
   * @param {Object} [options={}] - Execution options
   * @param {boolean} [options.changes=true] - Include changes in output
   * @param {boolean} [options.after=true] - Include state after changes
   * @param {boolean} [options.before=false] - Include state before changes
   * @returns {Promise<{stdout: string, stderr: string, exitCode: number}>}
   * @throws {Error} If execution fails or times out
   *
   * @example
   * // Create a link
   * const result = await process.execute('() ((1 2))');
   *
   * // Read all links
   * const links = await process.execute('((($i: $s $t)) (($i: $s $t)))');
   */
  execute(query, options = {}) {
    return new Promise((resolve, reject) => {
      const args = [
        '--db',
        this._options.dbPath,
        query,
        ...(options.changes !== false ? ['--changes'] : []),
        ...(options.after !== false ? ['--after'] : []),
        ...(options.before === true ? ['--before'] : []),
        ...(this._options.trace ? ['--trace'] : []),
      ];

      let stdout = '';
      let stderr = '';
      let timedOut = false;

      const proc = spawn(this._options.binaryPath, args, {
        stdio: ['ignore', 'pipe', 'pipe'],
      });

      const timeout = setTimeout(() => {
        timedOut = true;
        proc.kill('SIGTERM');
        reject(
          new Error(
            `Operation timed out after ${this._options.operationTimeout}ms`
          )
        );
      }, this._options.operationTimeout);

      proc.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      proc.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      proc.on('error', (error) => {
        clearTimeout(timeout);
        this.emit('error', error);
        reject(new Error(`Failed to execute link-cli: ${error.message}`));
      });

      proc.on('close', (exitCode) => {
        clearTimeout(timeout);
        if (timedOut) {
          return;
        }

        if (exitCode !== 0) {
          const errorMsg = stderr || `link-cli exited with code ${exitCode}`;
          reject(new Error(errorMsg));
          return;
        }

        resolve({ stdout, stderr, exitCode });
      });
    });
  }

  /**
   * Creates a link in the database.
   *
   * @param {number|string} source - Source reference
   * @param {number|string} target - Target reference
   * @returns {Promise<{id: number, source: number, target: number}|null>}
   *
   * @example
   * const link = await process.createLink(1, 2);
   * console.log(link); // { id: 1, source: 1, target: 2 }
   */
  async createLink(source, target) {
    const query = `() ((${source} ${target}))`;
    const result = await this.execute(query);

    // Parse the output to get the created link
    const created = this._parseChanges(result.stdout);
    if (created.length > 0) {
      return created[0];
    }
    return null;
  }

  /**
   * Reads all links matching a pattern.
   *
   * @param {Object} [pattern={}] - Pattern to match
   * @param {number|string} [pattern.id] - Link ID to match
   * @param {number|string} [pattern.source] - Source to match
   * @param {number|string} [pattern.target] - Target to match
   * @returns {Promise<Array<{id: number, source: number, target: number}>>}
   *
   * @example
   * // Read all links
   * const links = await process.readLinks();
   *
   * // Read links with specific source
   * const filtered = await process.readLinks({ source: 1 });
   */
  async readLinks(pattern = {}) {
    // Build the query based on pattern
    let query;
    if (pattern.id !== undefined) {
      // Specific ID lookup
      const s = pattern.source !== undefined ? pattern.source : '$s';
      const t = pattern.target !== undefined ? pattern.target : '$t';
      query = `(((${pattern.id}: ${s} ${t})) ((${pattern.id}: ${s} ${t})))`;
    } else if (pattern.source !== undefined || pattern.target !== undefined) {
      // Pattern-based query
      const s = pattern.source !== undefined ? pattern.source : '$s';
      const t = pattern.target !== undefined ? pattern.target : '$t';
      query = `((($i: ${s} ${t})) (($i: ${s} ${t})))`;
    } else {
      // Read all links
      query = '((($i: $s $t)) (($i: $s $t)))';
    }

    const result = await this.execute(query, { changes: true, after: false });

    // Parse the output to get the links
    return this._parseChanges(result.stdout);
  }

  /**
   * Deletes a link by its structure (source, target).
   *
   * @param {number|string} source - Source reference
   * @param {number|string} target - Target reference
   * @returns {Promise<boolean>} True if link was deleted
   *
   * @example
   * const deleted = await process.deleteLink(1, 2);
   */
  async deleteLink(source, target) {
    const query = `((${source} ${target})) ()`;
    const result = await this.execute(query);

    // Check if deletion happened
    const deleted = this._parseChanges(result.stdout);
    return deleted.length > 0;
  }

  /**
   * Deletes a link by its ID.
   *
   * @param {number|string} id - Link ID
   * @returns {Promise<boolean>} True if link was deleted
   *
   * @example
   * const deleted = await process.deleteLinkById(1);
   */
  async deleteLinkById(id) {
    // First, read the link to get its source and target
    const links = await this.readLinks({ id });
    if (links.length === 0) {
      return false;
    }

    const link = links[0];
    return this.deleteLink(link.source, link.target);
  }

  /**
   * Parses link-cli output to extract links.
   *
   * @private
   * @param {string} output - Raw output from link-cli
   * @returns {Array<{id: number, source: number, target: number}>}
   */
  _parseChanges(output) {
    const links = [];
    const trimmed = output.trim();

    if (!trimmed) {
      return links;
    }

    // Parse Links Notation format: ((id: source target) ...)
    // Match patterns like (1: 1 2) or (1 2)
    const linkPattern = /\((\d+):\s*(\d+)\s+(\d+)\)/g;
    let match;

    while ((match = linkPattern.exec(trimmed)) !== null) {
      links.push({
        id: parseInt(match[1], 10),
        source: parseInt(match[2], 10),
        target: parseInt(match[3], 10),
      });
    }

    // Also try to match simple format (source target) when id is implicit
    // This is used in --changes output
    const simplePattern = /\((\d+)\s+(\d+)\)/g;
    while ((match = simplePattern.exec(trimmed)) !== null) {
      // Check if this wasn't already matched as part of id:source target
      const fullCheck = new RegExp(`\\(\\d+:\\s*${match[1]}\\s+${match[2]}\\)`);
      if (!fullCheck.test(trimmed)) {
        // For simple format without ID, we'll need context to determine ID
        // For now, skip these as they're typically part of the query, not results
      }
    }

    return links;
  }
}

File: rust/src/backends/link_cli_process.rs
//! Process management for link-cli backend.
//!
//! This module handles spawning, communicating with, and managing the lifecycle
//! of link-cli child processes.
//!
//! # Overview
//!
//! The `LinkCliProcess` struct manages:
//! - Spawning the link-cli (clink) process
//! - Sending commands via stdin
//! - Reading responses from stdout
//! - Handling process crashes and restarts
//! - Graceful shutdown
//!
//! # Example
//!
//! ```rust,ignore
//! use links_queue::backends::link_cli_process::{LinkCliProcess, ProcessConfig};
//!
//! let config = ProcessConfig::new("./data/db.links");
//! let mut process = LinkCliProcess::new(config);
//!
//! process.start().await?;
//! let response = process.execute("() ((1 2))").await?;
//! process.stop().await?;
//! ```

use std::io;
use std::path::PathBuf;
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, Command};
use tokio::sync::Mutex;
use tokio::time::{timeout, Duration};

// =============================================================================
// Process Configuration
// =============================================================================

/// Configuration for the link-cli process.
#[derive(Debug, Clone)]
pub struct ProcessConfig {
    /// Path to the database file.
    pub db_path: PathBuf,

    /// Path to the link-cli binary (defaults to "clink" in PATH).
    pub binary_path: Option<PathBuf>,

    /// Timeout for command execution in milliseconds.
    pub command_timeout_ms: u64,

    /// Whether to auto-restart on crash.
    pub auto_restart: bool,

    /// Maximum restart attempts before giving up.
    pub max_restart_attempts: u32,
}

impl ProcessConfig {
    /// Creates a new process configuration with the given database path.
    ///
    /// # Arguments
    ///
    /// * `db_path` - Path to the database file
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use links_queue::backends::link_cli_process::ProcessConfig;
    ///
    /// let config = ProcessConfig::new("./data/db.links");
    /// ```
    #[must_use]
    pub fn new(db_path: impl Into<PathBuf>) -> Self {
        Self {
            db_path: db_path.into(),
            binary_path: None,
            command_timeout_ms: 30_000, // 30 seconds default
            auto_restart: true,
            max_restart_attempts: 3,
        }
    }

    /// Sets a custom binary path for link-cli.
    ///
    /// # Arguments
    ///
    /// * `path` - Path to the link-cli binary
    #[must_use]
    pub fn with_binary_path(mut self, path: impl Into<PathBuf>) -> Self {
        self.binary_path = Some(path.into());
        self
    }

    /// Sets the command execution timeout.
    ///
    /// # Arguments
    ///
    /// * `timeout_ms` - Timeout in milliseconds
    #[must_use]
    pub const fn with_timeout(mut self, timeout_ms: u64) -> Self {
        self.command_timeout_ms = timeout_ms;
        self
    }

    /// Sets whether to auto-restart on crash.
    ///
    /// # Arguments
    ///
    /// * `auto_restart` - Whether to auto-restart
    #[must_use]
    pub const fn with_auto_restart(mut self, auto_restart: bool) -> Self {
        self.auto_restart = auto_restart;
        self
    }

    /// Gets the binary path, defaulting to "clink".
    fn binary(&self) -> &str {
        self.binary_path
            .as_ref()
            .and_then(|p| p.to_str())
            .unwrap_or("clink")
    }
}

impl Default for ProcessConfig {
    fn default() -> Self {
        Self::new("db.links")
    }
}

// =============================================================================
// Process Error
// =============================================================================

/// Error that can occur during process operations.
#[derive(Debug)]
pub enum ProcessError {
    /// Failed to spawn the process.
    SpawnFailed(io::Error),
    /// Process is not running.
    NotRunning,
    /// Failed to write to stdin.
    WriteFailed(io::Error),
    /// Failed to read from stdout.
    ReadFailed(io::Error),
    /// Command execution timed out.
    Timeout,
    /// Process crashed unexpectedly.
    ProcessCrashed(Option<i32>),
    /// Maximum restart attempts exceeded.
    MaxRestartsExceeded,
    /// The binary was not found.
    BinaryNotFound(String),
}

impl std::fmt::Display for ProcessError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::SpawnFailed(e) => write!(f, "Failed to spawn link-cli process: {e}"),
            Self::NotRunning => write!(f, "Link-cli process is not running"),
            Self::WriteFailed(e) => write!(f, "Failed to write to link-cli stdin: {e}"),
            Self::ReadFailed(e) => write!(f, "Failed to read from link-cli stdout: {e}"),
            Self::Timeout => write!(f, "Command execution timed out"),
            Self::ProcessCrashed(code) => {
                if let Some(code) = code {
                    write!(f, "Link-cli process crashed with exit code {code}")
                } else {
                    write!(f, "Link-cli process crashed")
                }
            }
            Self::MaxRestartsExceeded => write!(f, "Maximum restart attempts exceeded"),
            Self::BinaryNotFound(path) => write!(f, "Link-cli binary not found: {path}"),
        }
    }
}

impl std::error::Error for ProcessError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::SpawnFailed(e) | Self::WriteFailed(e) | Self::ReadFailed(e) => Some(e),
            _ => None,
        }
    }
}

/// Result type for process operations.
pub type ProcessResult<T> = Result<T, ProcessError>;

// =============================================================================
// Process State
// =============================================================================

/// Internal state of a running process.
struct ProcessState {
    /// The child process handle.
    child: Child,
    /// Buffered reader for stdout.
    stdout_reader: BufReader<tokio::process::ChildStdout>,
    /// Writer for stdin.
    stdin_writer: tokio::process::ChildStdin,
}

// =============================================================================
// Link CLI Process
// =============================================================================

/// Manager for a link-cli child process.
///
/// Handles process lifecycle, command execution, and error recovery.
pub struct LinkCliProcess {
    /// Process configuration.
    config: ProcessConfig,
    /// Current process state (None if not running).
    state: Mutex<Option<ProcessState>>,
    /// Number of restart attempts.
    restart_count: Mutex<u32>,
}

impl LinkCliProcess {
    /// Creates a new process manager with the given configuration.
    ///
    /// # Arguments
    ///
    /// * `config` - Process configuration
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use links_queue::backends::link_cli_process::{LinkCliProcess, ProcessConfig};
    ///
    /// let config = ProcessConfig::new("./data/db.links");
    /// let process = LinkCliProcess::new(config);
    /// ```
    #[must_use]
    pub fn new(config: ProcessConfig) -> Self {
        Self {
            config,
            state: Mutex::new(None),
            restart_count: Mutex::new(0),
        }
    }

    /// Checks if the process is currently running.
    pub async fn is_running(&self) -> bool {
        let state = self.state.lock().await;
        state.is_some()
    }

    /// Starts the link-cli process.
    ///
    /// # Errors
    ///
    /// Returns `ProcessError::SpawnFailed` if the process cannot be started.
    /// Returns `ProcessError::BinaryNotFound` if clink is not installed.
    #[allow(clippy::significant_drop_tightening)]
    pub async fn start(&self) -> ProcessResult<()> {
        let mut state = self.state.lock().await;

        // If already running, do nothing
        if state.is_some() {
            return Ok(());
        }

        let process_state = self.spawn_process()?;
        *state = Some(process_state);
        *self.restart_count.lock().await = 0;

        Ok(())
    }

    /// Spawns a new link-cli process.
    fn spawn_process(&self) -> ProcessResult<ProcessState> {
        let binary = self.config.binary();
        let db_path = self.config.db_path.to_string_lossy();

        let mut child = Command::new(binary)
            .arg("--db")
            .arg(db_path.as_ref())
            .arg("--interactive")
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .spawn()
            .map_err(|e| {
                if e.kind() == io::ErrorKind::NotFound {
                    ProcessError::BinaryNotFound(binary.to_string())
                } else {
                    ProcessError::SpawnFailed(e)
                }
            })?;

        let stdin = child.stdin.take().ok_or_else(|| {
            ProcessError::SpawnFailed(io::Error::new(
                io::ErrorKind::Other,
                "Failed to capture stdin",
            ))
        })?;

        let stdout = child.stdout.take().ok_or_else(|| {
            ProcessError::SpawnFailed(io::Error::new(
                io::ErrorKind::Other,
                "Failed to capture stdout",
            ))
        })?;

        Ok(ProcessState {
            child,
            stdout_reader: BufReader::new(stdout),
            stdin_writer: stdin,
        })
    }

    /// Stops the link-cli process gracefully.
    ///
    /// # Errors
    ///
    /// Returns an error if the process cannot be stopped.
    #[allow(clippy::significant_drop_tightening)]
    pub async fn stop(&self) -> ProcessResult<()> {
        let mut state = self.state.lock().await;

        if let Some(mut process_state) = state.take() {
            // Try to kill gracefully
            let _ = process_state.child.kill().await;
        }

        Ok(())
    }

    /// Executes a command and returns the response.
    ///
    /// # Arguments
    ///
    /// * `command` - The Links Notation command to execute
    ///
    /// # Returns
    ///
    /// The response from link-cli (may be empty for some operations).
    ///
    /// # Errors
    ///
    /// Returns an error if the command fails or times out.
    pub async fn execute(&self, command: &str) -> ProcessResult<String> {
        self.execute_with_options(command, true).await
    }

    /// Executes a command with options.
    ///
    /// # Arguments
    ///
    /// * `command` - The Links Notation command to execute
    /// * `expect_response` - Whether to wait for a response
    #[allow(clippy::significant_drop_tightening)]
    async fn execute_with_options(
        &self,
        command: &str,
        expect_response: bool,
    ) -> ProcessResult<String> {
        let mut state_guard = self.state.lock().await;

        let state = state_guard.as_mut().ok_or(ProcessError::NotRunning)?;

        // Write command to stdin
        let cmd_with_newline = format!("{command}\n");
        state
            .stdin_writer
            .write_all(cmd_with_newline.as_bytes())
            .await
            .map_err(ProcessError::WriteFailed)?;

        state
            .stdin_writer
            .flush()
            .await
            .map_err(ProcessError::WriteFailed)?;

        if !expect_response {
            return Ok(String::new());
        }

        // Read response with timeout
        let timeout_duration = Duration::from_millis(self.config.command_timeout_ms);
        let mut response = String::new();

        match timeout(
            timeout_duration,
            state.stdout_reader.read_line(&mut response),
        )
        .await
        {
            Ok(Ok(_)) => Ok(response.trim().to_string()),
            Ok(Err(e)) => Err(ProcessError::ReadFailed(e)),
            Err(_) => Err(ProcessError::Timeout),
        }
    }

    /// Attempts to restart the process after a failure.
    ///
    /// # Errors
    ///
    /// Returns `ProcessError::MaxRestartsExceeded` if too many restarts.
    pub async fn try_restart(&self) -> ProcessResult<()> {
        if !self.config.auto_restart {
            return Err(ProcessError::NotRunning);
        }

        let mut count = self.restart_count.lock().await;
        if *count >= self.config.max_restart_attempts {
            return Err(ProcessError::MaxRestartsExceeded);
        }

        *count += 1;
        drop(count);

        // Stop existing process if any
        self.stop().await?;

        // Start new process
        self.start().await
    }

    /// Gets the current configuration.
    #[must_use]
    pub const fn config(&self) -> &ProcessConfig {
        &self.config
    }
}

impl Drop for LinkCliProcess {
    fn drop(&mut self) {
        // Best effort cleanup - can't await in drop
        if let Ok(mut state) = self.state.try_lock() {
            if let Some(mut process_state) = state.take() {
                // Try to kill the process synchronously
                let _ = process_state.child.start_kill();
            }
        }
    }
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    mod config_tests {
        use super::*;

        #[test]
        fn test_config_new() {
            let config = ProcessConfig::new("test.links");
            assert_eq!(config.db_path, PathBuf::from("test.links"));
            assert!(config.binary_path.is_none());
            assert_eq!(config.command_timeout_ms, 30_000);
            assert!(config.auto_restart);
        }

        #[test]
        fn test_config_builder() {
            let config = ProcessConfig::new("test.links")
                .with_binary_path("/usr/bin/clink")
                .with_timeout(5000)
                .with_auto_restart(false);

            assert_eq!(config.binary_path, Some(PathBuf::from("/usr/bin/clink")));
            assert_eq!(config.command_timeout_ms, 5000);
            assert!(!config.auto_restart);
        }

        #[test]
        fn test_config_default() {
            let config = ProcessConfig::default();
            assert_eq!(config.db_path, PathBuf::from("db.links"));
        }

        #[test]
        fn test_config_binary_default() {
            let config = ProcessConfig::new("test.links");
            assert_eq!(config.binary(), "clink");
        }

        #[test]
        fn test_config_binary_custom() {
            let config = ProcessConfig::new("test.links").with_binary_path("/custom/clink");
            assert_eq!(config.binary(), "/custom/clink");
        }
    }

    mod error_tests {
        use super::*;

        #[test]
        fn test_error_display() {
            let err = ProcessError::NotRunning;
            assert!(err.to_string().contains("not running"));

            let err = ProcessError::Timeout;
            assert!(err.to_string().contains("timed out"));

            let err = ProcessError::ProcessCrashed(Some(1));
            assert!(err.to_string().contains("exit code 1"));

            let err = ProcessError::BinaryNotFound("clink".to_string());
            assert!(err.to_string().contains("clink"));
        }
    }

    mod process_tests {
        use super::*;

        #[tokio::test]
        async fn test_process_new() {
            let config = ProcessConfig::new("test.links");
            let process = LinkCliProcess::new(config);
            assert!(!process.is_running().await);
        }

        #[tokio::test]
        async fn test_process_not_running_error() {
            let config = ProcessConfig::new("test.links");
            let process = LinkCliProcess::new(config);

            let result = process.execute("() ((1 2))").await;
            assert!(matches!(result, Err(ProcessError::NotRunning)));
        }

        #[tokio::test]
        async fn test_start_without_binary() {
            let config = ProcessConfig::new("test.links").with_binary_path("/nonexistent/clink");
            let process = LinkCliProcess::new(config);

            let result = process.start().await;
            assert!(result.is_err());
        }
    }
}
